samples/emoji-search/android-composeui/src/main/java/app/cash/zipline/samples/emojisearch/composeui/ComposeUiTextInput.kt

```diff
@@ -15,44 +15,91 @@
  */
 package app.cash.zipline.samples.emojisearch.composeui
 
-import androidx.compose.foundation.layout.fillMaxWidth
-import androidx.compose.foundation.layout.padding
 import androidx.compose.material.Text
 import androidx.compose.material.TextField
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.getValue
 import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.setValue
-import androidx.compose.ui.Modifier
-import androidx.compose.ui.unit.dp
+import androidx.compose.ui.text.TextRange
+import androidx.compose.ui.text.input.TextFieldValue
 import app.cash.redwood.LayoutModifier
+import app.cash.redwood.treehouse.TreehouseDispatchers
 import example.schema.widget.TextInput
+import example.values.TextFieldState
+import kotlin.coroutines.EmptyCoroutineContext
 
-class ComposeUiTextInput : TextInput<@Composable () -> Unit> {
+class ComposeUiTextInput(
+  private val dispatchers: TreehouseDispatchers,
+) : TextInput<@Composable () -> Unit> {
+  private var state by mutableStateOf(TextFieldState())
   private var hint by mutableStateOf("")
-  private var text by mutableStateOf("")
-  private var onTextChanged: ((String) -> Unit)? = null
+  private var onChange: ((TextFieldState) -> Unit)? = null
+  private var updating = false
 
   override var layoutModifiers: LayoutModifier = LayoutModifier
 
   override val value = @Composable {
     TextField(
-      value = text,
-      onValueChange = { onTextChanged?.invoke(it) },
-      label = { Text(hint) },
+      value = TextFieldValue(
+        text = state.text,
+        selection = TextRange(state.selectionStart, state.selectionEnd)
+      ),
```

Comment:
`CoreTextField` will call `onValueChange` when the IME updates the `composition` property on `TextFieldValue`. Preserving that value is why I was keeping track of the `TextFieldValue` coming back from `onValueChanged` but that also led to some more tricky state handling.

Reply:
Running the sample on this branch shows that the `composition` is getting dropped (e.g. the IME underlines the text that is being typed). I don't know if that's a big deal or not, but I thought it would be good to try and keep it.

Reply:
@eygraber Great callout - I added a commit to remember the latest `TextFieldValue` to preserve that state without recomposing.

Reply:
This causes an infinite loop of `onValueChange` being called when any text is typed.

Reply:
In fact reverting `1f53621b7d` will lead to the same infinite loop of `onValueChange` invocations.

Reply:
@eygraber Ah yep, you're right as far as I can tell `onValueChange` loops forever if you don't use its value. We could probably create some higher order state object that combines the values from `TextFieldValue` and `TextFieldState`, but that's probably overkill. Since we use `textFieldValue.copy` the `text` and `selection` values on `TextFieldValue` are always ignored.

Reply:
:+1: 

In theory there can be a case where the presenter mutates the text under certain conditions that can be skipped if the user edit takes precedence over the Treehouse update (as a contrived example, a presenter that truncates the text to 9 characters if the text is 10 characters long). Can't think of any real world examples at the moment, and also having trouble deciding if that's a bug or just a consequence of text input :man_shrugging: